core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java

```diff
@@ -385,29 +405,427 @@ public synchronized void close()
+        public Slice getNext()
+        {
+            if (!iterator.hasNext()) {
+                return null;
+            }
+            return iterator.next();
+        }
+
+        @Override
+        public boolean isFinished()
+        {
+            return !iterator.hasNext();
+        }
+
+        @Override
+        public ListenableFuture<Void> isBlocked()
+        {
+            return immediateVoidFuture();
+        }
+
+        @Override
+        public long getRetainedSizeInBytes()
+        {
+            return 0;
+        }
+
+        @Override
+        public void close()
+        {
+        }
+    }
+
+    @NotThreadSafe
+    private static class ExchangeOutputSource
+            implements OutputSource
+    {
+        private final Set<TaskId> selectedTasks;
+        private final ExchangeManager exchangeManager;
+        private final Exchange exchange;
+        private final QueryId queryId;
+
+        private ExchangeSource exchangeSource;
+        private boolean finished;
+
+        private ExchangeOutputSource(
+                Set<TaskId> selectedTasks,
+                ExchangeManager exchangeManager,
+                Exchange exchange,
+                QueryId queryId)
+        {
+            this.selectedTasks = ImmutableSet.copyOf(requireNonNull(selectedTasks, "selectedTasks is null"));
+            this.exchangeManager = requireNonNull(exchangeManager, "exchangeManager is null");
+            this.exchange = requireNonNull(exchange, "exchange is null");
+            this.queryId = requireNonNull(queryId, "queryId is null");
+        }
+
+        @Override
+        public Slice getNext()
+        {
+            if (finished) {
+                return null;
+            }
+            if (exchangeSource == null) {
+                CompletableFuture<List<ExchangeSourceHandle>> sourceHandlesFuture = exchange.getSourceHandles();
+                if (!sourceHandlesFuture.isDone()) {
+                    return null;
+                }
+                List<ExchangeSourceHandle> handles = getUnchecked(sourceHandlesFuture);
+                exchangeSource = exchangeManager.createSource(handles);
+            }
+            while (!exchangeSource.isFinished()) {
+                if (!exchangeSource.isBlocked().isDone()) {
+                    return null;
+                }
+                Slice buffer = exchangeSource.read();
+                if (buffer == null) {
+                    continue;
+                }
+                int stageId = buffer.getInt(0);
+                int partitionId = buffer.getInt(Integer.BYTES);
+                int attemptId = buffer.getInt(Integer.BYTES * 2);
+                TaskId taskId = new TaskId(new StageId(queryId, stageId), partitionId, attemptId);
+                if (!selectedTasks.contains(taskId)) {
+                    continue;
+                }
+                return buffer.slice(Integer.BYTES * 3, buffer.length() - Integer.BYTES * 3);
+            }
+            close();
```

Comment:
It is to have only the `finished` variable instead of both `finished` and `closed` that seems redundant. It set's `finished` to `true` and closes the source. It is also correct to set `finished` to true here directly, but then the source should be closed in `close` and be controlled by a second `closed` variable.